### PR TITLE
feature(events): add hashChangeStart and hashChangeComplete

### PR DESCRIPTION
**Pull request by jantimon**
this PR adds support for two router events (see [https://nextjs.org/docs/api-reference/next/router#routerevents](https://nextjs.org/docs/api-reference/next/router#routerevents)):

- hashChangeStart
- hashChangeComplete

it uses the current url of the MemoryRouter to detect if the `hash` or the `route` was changed and triggers only the matching events (that's the behaviour of next 11)

unfortunately the unit will now test fail if the previous test navigated to the same route.  
I adjusted the test (second commit) to go to a dedicated route instead but I am not sure if this is the desired solution

### CODE REVIEW

File: src/MemoryRouter.tsx [hunk 1]
```diff
@@ -145,9 +146,20 @@ export class MemoryRouter extends BaseRouter {
...
+    const isHashChange = this.hash !== hash;
+    const isQueryChange = stringifyQueryString(this.query) !== stringifyQueryString(query);
+    const isRouteChange = isQueryChange || this.pathname !== pathname;
+
+    if (isHashChange) {
+      this.events.emit("hashChangeStart", asPath, { shallow });
+    }
+
+    if (isRouteChange) {
+      this.events.emit("routeChangeStart", asPath, { shallow });
```
Comment (scottrippey) [inline src/MemoryRouter.tsx:161]: In a Next app, if I was to call `router.push(window.location)` (ie. the path doesn't change at all), any idea if it would still fire the `routeChangeStart/Complete` events?

File: src/useMemoryRouter.tsx [hunk 2]
```diff
@@ -34,23 +36,29 @@ export const useMemoryRouter = (singletonRouter: MemoryRouter, eventHandlers?: M
   // Subscribe to any eventHandlers:
   useEffect(() => {
     if (!eventHandlers) return;
-    const { onRouteChangeStart, onRouteChangeComplete, onPush, onReplace } = eventHandlers;
-    if (!(onRouteChangeStart || onRouteChangeComplete || onPush || onReplace)) return;
+    const { onRouteChangeStart, onRouteChangeComplete, onHashChangeComplete, onHashChangeStart, onPush, onReplace } = eventHandlers;
+    if (!(onRouteChangeStart || onRouteChangeComplete  || onHashChangeStart ||onHashChangeComplete || onPush || onReplace)) return;
```
Comment (scottrippey) [inline src/useMemoryRouter.tsx:40]: Minor nit:
```suggestion
    if (!(onRouteChangeStart || onRouteChangeComplete  || onHashChangeStart || onHashChangeComplete || onPush || onReplace)) return;
```
I should add precommit hooks or something, this should be automated

File: src/useMemoryRouter.tsx [hunk 3]
```diff
@@ -34,23 +36,29 @@ export const useMemoryRouter = (singletonRouter: MemoryRouter, eventHandlers?: M
...
 
     if (onRouteChangeStart) singletonRouter.events.on("routeChangeStart", onRouteChangeStart);
     if (onRouteChangeComplete) singletonRouter.events.on("routeChangeComplete", onRouteChangeComplete);
+    if (onHashChangeStart) singletonRouter.events.on("hashChangeStart", onHashChangeStart);
+    if (onRouteChangeComplete) singletonRouter.events.on("hashChangeComplete", onRouteChangeComplete);
```
Comment (scottrippey) [inline src/useMemoryRouter.tsx:45]: This looks like a typo
```suggestion
    if (onHashChangeComplete) singletonRouter.events.on("hashChangeComplete", onHashChangeComplete);
```

**Review by scottrippey — CHANGES_REQUESTED** [review]
Looks great overall.  One major typo to fix, some other tiny comments.

**Comment by jantimon** [comment]
Thanks for your feedback :)

I'll fix the code as you suggested.

I also created a codesandbox to test Next Router Events:
https://codesandbox.io/s/nextjs-routerevent-tests-fwtzu

The results are the following:

```jsx
// Same route will trigger routeChangeStart and routeChange complete
router.push("/") 
    -> routeChangeStart 
    -> routeChangeComplete 
router.push("/") 
    -> routeChangeStart 
    -> routeChangeComplete 

// Different route will trigger routeChangeStart and routeChange complete
router.push("/") 
    -> routeChangeStart 
    -> routeChangeComplete 
router.push("/foo") 
    -> routeChangeStart 
    -> routeChangeComplete 


// Changing only the hash will trigger hashChangeStart and hashChangeComplete
router.push("/foo") 
    -> routeChangeStart 
    -> routeChangeComplete 
router.push("/foo#baz") 
    -> hashChangeStart 
    -> hashChangeComplete 


// Changing hash and route will trigger only routeChangeStart and routeChangeComplete
router.push("/") 
    -> routeChangeStart 
    -> routeChangeComplete 
router.push("/foo#baz") 
    -> routeChangeStart 
    -> routeChangeComplete 
```

I'll adjust the behaviour for the memory router accordingly

File: src/MemoryRouter.tsx (hunk 1 above)
Comment (jantimon) [reply src/MemoryRouter.tsx:160]: yes you can test it here: https://codesandbox.io/s/nextjs-routerevent-tests-fwtzu

File: src/useMemoryRouter.tsx (hunk 2 above)
Comment (jantimon) [reply src/useMemoryRouter.tsx]: I removed this if statement and ran prettier

File: src/useMemoryRouter.tsx (hunk 3 above)
Comment (jantimon) [reply src/useMemoryRouter.tsx]: oh 😱 - nice find

File: src/MemoryRouter.test.tsx [hunk 4]
```diff
@@ -62,6 +68,19 @@ describe("MemoryRouter", () => {
           });
         });
 
+        it("should trigger only hashEvents on a hash change event", async () => {
+          routeChangeStart.mockClear();
+          routeChangeComplete.mockClear();
+          hashChangeStart.mockClear();
+          hashChangeComplete.mockClear();
+
```
Comment (scottrippey) [inline src/MemoryRouter.test.tsx:76]: There's already a `beforeEach` that calls `jest.clearAllMocks()`
```suggestion
```

File: src/MemoryRouter.tsx [hunk 5]
```diff
@@ -145,21 +146,36 @@ export class MemoryRouter extends BaseRouter {
...
-    const asPath = getRouteAsPath(baseUrlObject.pathname ?? "", baseQuery, urlObject.hash);
-
-    this.events.emit("routeChangeStart", asPath, { shallow });
+    const hash = urlObject.hash || "";
+    const asPath = getRouteAsPath(baseUrlObject.pathname ?? "", baseQuery, hash);
+
+    const isHashChange = this.hash !== hash;
+    const isQueryChange = stringifyQueryString(this.query) !== stringifyQueryString(query);
+    const isRouteChange = isQueryChange || this.pathname !== pathname;
+    const isOnlyHashChange = isHashChange && !isRouteChange;
```
Comment (scottrippey) [inline src/MemoryRouter.tsx:155]: Thank you so much for the codesandbox (I forgot that they supported Next apps, because otherwise it's tedious to set up my own testing!).
Playing with it, I found there was a little bit of edge-case behavior, and perhaps we should try to capture that logic.
The edge-case is when I call `push('/foo#bar')` twice.  In Next, this triggers the hash-only change, even though the hash did NOT change.  So, to be as accurate as possible, I think the logic should be:

```suggestion
    const isQueryChange = stringifyQueryString(this.query) !== stringifyQueryString(query);
    const isRouteChange = isQueryChange || this.pathname !== pathname;
    const isOnlyHashChange = !isRouteChange && hash;
```

File: src/MemoryRouter.test.tsx [hunk 6]
```diff
@@ -62,6 +68,19 @@ describe("MemoryRouter", () => {
           });
         });
 
+        it("should trigger only hashEvents on a hash change event", async () => {
```
Comment (scottrippey) [inline src/MemoryRouter.test.tsx:71]: I would love a few more test cases to ensure we capture the Next behavior.

File: src/MemoryRouter.tsx (hunk 5 above)
Comment (jantimon) [reply src/MemoryRouter.tsx:155]: Unfortunately there is one more edge case - removing the hash (e.g. `/foo#hash` -> `/foo`) will also trigger the hashChange:

```
/foo
  -> routeChange
/foo#baz
  -> hashChange
/foo#baz
  -> hashChange
/foo
  -> hashChange
/foo
  -> routeChange
```

File: src/MemoryRouter.test.tsx (hunk 6 above)
Comment (jantimon) [reply src/MemoryRouter.test.tsx]: I added unit tests for (hopefully) all cases

File: src/MemoryRouter.tsx (hunk 5 above)
Comment (jantimon) [reply src/MemoryRouter.tsx]: I added a fix and unit tests for all those cases

File: src/MemoryRouter.test.tsx (hunk 4 above)
Comment (jantimon) [reply src/MemoryRouter.test.tsx]: Okay - I'll fix that

File: src/MemoryRouter.test.tsx [hunk 7]
```diff
@@ -68,19 +74,56 @@ describe("MemoryRouter", () => {
...
+
+        it("should trigger only routeEvents for /baz -> /foo#baz", async () => {
+          await memoryRouter.push("/baz");
+          clearListenerMocks();
+          await memoryRouter.push("/foo#baz");
+          expect(hashChangeStart).not.toHaveBeenCalled();
+          expect(hashChangeComplete).not.toHaveBeenCalled();
+          expect(routeChangeStart).toHaveBeenCalledWith("/foo#baz", { shallow: false });
+          expect(routeChangeComplete).toHaveBeenCalledWith("/foo#baz", { shallow: false });
+        });
```
Comment (scottrippey) [inline src/MemoryRouter.test.tsx:125]: These tests are amazing, well done!

**Comment by scottrippey** [comment]
Thanks for sticking with it, this PR is perfect.  Will publish soon!

**Review by scottrippey — APPROVED** [review]

**Comment by scottrippey** [comment]
Published as version `0.6.4`!  
Thank you for implementing this.

**Comment by jantimon** [comment]
Thank you so much for your time and for publishing it! :)
